src/torrent.cpp

```diff
@@ -1384,16 +1382,89 @@ bool is_downloading_state(int const st)
 #endif
 
 #ifdef TORRENT_SSL_PEERS
+	void torrent::init_ssl(string_view cert)
+	{
+		// create the SSL context for this torrent. We need to
+		// inject the root certificate, and no other, to
+		// verify other peers against
+#if BOOST_VERSION >= 106400
+		std::unique_ptr<ssl::context> ctx(new ssl::context(ssl::context::tls));
+#else
+		std::unique_ptr<ssl::context> ctx(new ssl::context(ssl::context::tlsv12));
+#endif
+
+		ctx->set_options(ssl::context::default_workarounds
+			| ssl::context::no_sslv2
+			| ssl::context::no_sslv3
+			| ssl::context::single_dh_use);
+
+		error_code ec;
+		ctx->set_verify_mode(ssl::context::verify_peer
+			| ssl::context::verify_fail_if_no_peer_cert
+			| ssl::context::verify_client_once, ec);
+		if (ec)
+		{
+			set_error(ec, torrent_status::error_file_ssl_ctx);
+			pause();
+			return;
+		}
+
+		// the verification function verifies the distinguished name
+		// of a peer certificate to make sure it matches the info-hash
+		// of the torrent, or that it's a "star-cert"
+		ctx->set_verify_callback(
+				std::bind(&torrent::verify_peer_cert, this, m_torrent_file->name(), _1, _2)
+				, ec);
+		if (ec)
+		{
+			set_error(ec, torrent_status::error_file_ssl_ctx);
+			pause();
+			return;
+		}
+
+		// set the root certificate as trust
+		ssl::set_trust_certificate(ctx->native_handle(), cert, ec);
+		if (ec)
+		{
+			set_error(ec, torrent_status::error_file_ssl_ctx);
+			pause();
+			return;
+		}
+
+		// if all went well, set the torrent ssl context to this one
+		m_ssl_ctx = std::move(ctx);
+		// tell the client we need a cert for this torrent
+		alerts().emplace_alert<torrent_need_cert_alert>(get_handle());
+
+#if 0
+		char filename[100];
+		std::snprintf(filename, sizeof(filename), "/tmp/%u.pem", random());
+		FILE* f = fopen(filename, "w+");
+		fwrite(cert.c_str(), cert.size(), 1, f);
+		fclose(f);
+		ctx->load_verify_file(filename);
+#endif
+	}
 
-	bool torrent::verify_peer_cert(bool const preverified, boost::asio::ssl::verify_context& ctx)
+	bool torrent::verify_peer_cert(std::string const& expected, bool preverified, ssl::verify_context& ctx)
 	{
 		// if the cert wasn't signed by the correct CA, fail the verification
 		if (!preverified) return false;
 
+#ifndef TORRENT_DISABLE_LOGGING
+		std::string names;
+		bool match = false;
+#endif
```

Comment:
these 4 lines used to be right before the loop with `sk_GENERAL_NAME_num`. It seems unnecessary to make the patch larger by moving them

Reply:
I put them here rather than putting them twice, since they are shared between OpenSSL and GnuTLS.

Reply:
ah, of course. makes sense